gnc/perception/sub8_pcl/nodes/buoy.cc

```diff
@@ -0,0 +1,291 @@
+
+  // believe it or not, this is on purpose (So blue replaces red in HSV)
+  cv::cvtColor(image_raw, image_hsv, CV_RGB2HSV);
+
+  cv::Point2d pt_cv_2d(250, 250);
+  sub::PointXYZT pcl_pt_3d;
+  pcl_pt_3d = sub::project_uv_to_cloud(*point_cloud_raw, pt_cv_2d, camera_model);
+
+  // Reprojection (Useful for visualization)
+  cv::Point2d cv_pt_uv =
+      cam_model.project3dToPixel(cv::Point3f(pcl_pt_3d.x, pcl_pt_3d.y, pcl_pt_3d.z));
+
+  // Threshold -- > This is what must be replaced with better 2d vision
+  cv::inRange(image_hsv, cv::Scalar(105, 135, 135), cv::Scalar(120, 255, 255), image_thresh);
+  std::vector<sub::Contour> contours;
+  std::vector<cv::Vec4i> hierarchy;
+  cv::findContours(image_thresh.clone(), contours, hierarchy, CV_RETR_EXTERNAL,
+                   CV_CHAIN_APPROX_SIMPLE, cv::Point(0, 0));
+  // TODO: ^^^ Make into function ^^^
+
+  // Not returning segmented point cloud for now
+  // sub::PointCloudT::Ptr segmented_cloud(new sub::PointCloudT());
+  // Vector of Eigen-vectors
+  std::vector<Eigen::Vector3f, Eigen::aligned_allocator<Eigen::Vector3f> > buoy_centers;
+
+  // Loop through the contours we found and find the positions
+  for (size_t i = 0; i < contours.size(); i++) {
+    cv::Point contour_centroid = sub::contour_centroid(contours[i]);
+    std::vector<cv::Point> approx;
+
+    // Magic num: 5
+    cv::approxPolyDP(contours[i], approx, 5, true);
+    double area = cv::contourArea(approx);
+
+    // Ignore small regions
+    if (area < 100) {
+      continue;
+    }
+    sub::PointXYZT centroid_projected =
+        sub::project_uv_to_cloud(*point_cloud_raw, contour_centroid, camera_model);
+    buoy_centers.push_back(sub::point_to_eigen(centroid_projected));
+  }
+  // ^^^^ The above could be eliminated with better approach
+
+  // TODO: initialize to inf
+  double closest_distance = 1000;
+  size_t closest_buoy_index = 0;
+
+  for (size_t buoy_index = 0; buoy_index < buoy_centers.size(); buoy_index++) {
+    const double distance = buoy_centers[buoy_index].norm();
+    if (buoy_centers[buoy_index].norm() < closest_distance) {
+      closest_distance = distance;
+      closest_buoy_index = buoy_index;
+    }
+  }
+  center = buoy_centers[closest_buoy_index];
+}
+
+void Sub8BuoyDetector::image_callback(const sensor_msgs::ImageConstPtr &image_msg,
+                                      const sensor_msgs::CameraInfoConstPtr &info_msg) {
+  need_new_cloud = true;
+
+#ifdef VISUALIZE
+  pcl::console::print_highlight("Getting image\n");
+#endif
+
+  cv_bridge::CvImagePtr input_bridge;
+  try {
+    input_bridge = cv_bridge::toCvCopy(image_msg, sensor_msgs::image_encodings::BGR8);
+    current_image = input_bridge->image;
+  } catch (cv_bridge::Exception &ex) {
+    ROS_ERROR("[draw_frames] Failed to convert image");
+    return;
+  }
+  cam_model.fromCameraInfo(info_msg);
+  image_time = image_msg->header.stamp;
+}
+
+void Sub8BuoyDetector::cloud_callback(const sensor_msgs::PointCloud2::ConstPtr &input_cloud) {
```

Comment:
Again, it seems like we could skip point cloud processing in the absence of a VisionRequest service call.